File: tests/revocation_propagation.py
"""Python parity for the G-6 release test.

Revokes a parent grant and measures how long the child grant keeps being
allowed by ``enforce(revocation_check="feed")``. Runs against a real auth
service with Postgres and Redis behind it (scripts/revocation-release-test.sh),
never against production: the base URL must be given explicitly.

    REVOCATION_RELEASE_BASE_URL=http://127.0.0.1:3199 python tests/revocation_propagation.py
"""
from __future__ import annotations

import json
import math
import os
import sys
import time
from pathlib import Path
from typing import Any

import httpx

import grantex
from grantex import DenialReason, Grantex, Permission, RevocationSubReason, ToolManifest
from grantex._types import AuthorizeParams, ExchangeTokenParams, SignupParams

# The SDK under test must be the one in this checkout: an ambient install of
# `grantex` would happily "prove" the acceptance criterion against code that is
# not being reviewed.
_EXPECTED_SRC = (Path(__file__).resolve().parents[1] / "packages" / "sdk-py" / "src").resolve()
_LOADED = Path(grantex.__file__).resolve().parent.parent
if _LOADED != _EXPECTED_SRC:
    print(
        f"refusing to measure: grantex was imported from {_LOADED}, not {_EXPECTED_SRC}. "
        "Run with PYTHONPATH=packages/sdk-py/src, or install this checkout.",
        file=sys.stderr,
    )
    raise SystemExit(2)

BASE_URL = os.environ.get("REVOCATION_RELEASE_BASE_URL")
TRIALS = int(os.environ.get("REVOCATION_RELEASE_TRIALS", "40"))
WARMUP = int(os.environ.get("REVOCATION_RELEASE_WARMUP", "2"))
# The requirement (PRD G-6), deliberately not overridable.
BUDGET_MS = 2000.0
REPORT = os.environ.get("REVOCATION_RELEASE_REPORT_PY")
SCOPES = ["tool:acme_kyb:read"]


def _percentile(values: list[float], fraction: float) -> float:
    ordered = sorted(values)
    index = min(len(ordered) - 1, math.ceil(fraction * len(ordered)) - 1)
    return ordered[max(0, index)]


def _approve(api_key: str, auth_request_id: str) -> str:
    response = httpx.post(
        f"{BASE_URL}/v1/authorize/{auth_request_id}/approve",
        headers={"Authorization": f"Bearer {api_key}", "Content-Type": "application/json"},
        json={},
        timeout=30.0,
    )
    response.raise_for_status()
    return str(response.json()["code"])


def main() -> int:
    if not BASE_URL:
        print("REVOCATION_RELEASE_BASE_URL is not set; skipping", file=sys.stderr)
        return 0

    account = Grantex.signup(
        SignupParams(name=f"revocation-release-py-{int(time.time())}", mode="sandbox"),
        base_url=BASE_URL,
    )
    admin = Grantex(api_key=account.api_key, base_url=BASE_URL)
    stamp = int(time.time())
    root = admin.agents.register(name=f"release-py-root-{stamp}", scopes=SCOPES)
    middle = admin.agents.register(name=f"release-py-middle-{stamp}", scopes=SCOPES)
    leaf = admin.agents.register(name=f"release-py-leaf-{stamp}", scopes=SCOPES)

    auth = admin.authorize(
        AuthorizeParams(agent_id=root.id, user_id=f"release-py-user-{stamp}", scopes=SCOPES)
    )
    code = auth.code or _approve(account.api_key, auth.request_id)
    root_grant = admin.tokens.exchange(ExchangeTokenParams(code=code, agent_id=root.id))

    pairs: list[tuple[str, str]] = []
    for _ in range(TRIALS + WARMUP):
        parent = admin.grants.delegate(
            parent_grant_token=root_grant.grant_token, sub_agent_id=middle.id, scopes=SCOPES, expires_in="1h"
        )
        child = admin.grants.delegate(
            parent_grant_token=parent["grantToken"], sub_agent_id=leaf.id, scopes=SCOPES, expires_in="1h"
        )
        pairs.append((parent["grantId"], child["grantToken"]))

    enforcer = Grantex(
        api_key=account.api_key,
        base_url=BASE_URL,
        revocation_check="feed",
        revocation_feed_stale_after=5.0,
    )
    enforcer.load_manifest(
        ToolManifest(connector="acme_kyb", tools={"resolve_business": Permission.READ})
    )
    try:
        if not enforcer.revocation_feed().ready(timeout=10.0):
            print("the revocation feed never became fresh", file=sys.stderr)
            return 1

        allowed = enforcer.enforce(pairs[0][1], "acme_kyb", "resolve_business")
        if not allowed.allowed:
            print(f"the child grant was denied before any revocation: {allowed.reason}", file=sys.stderr)
            return 1

        latencies: list[float] = []
        revoke_calls: list[float] = []
        failures = 0
        for index, (parent_grant_id, child_token) in enumerate(pairs):
            started = time.time()
            admin.grants.revoke(parent_grant_id)
            # Propagation is measured from the moment the revocation is
            # committed - when the API returns - not from when the call was
            # made. A developer on the free plan is rate limited to 100
            # requests a minute and the SDK waits out ``Retry-After``; that
            # wait is reported separately rather than charged to the feed.
            revoked_at = time.time()
            revoke_ms = (revoked_at - started) * 1000
            if revoke_ms > 1000:
                print(f"trial {index}: the revoke call waited {revoke_ms:.0f} ms (plan rate limit), not counted")
            denial: Any = None
            while time.time() - revoked_at < 30:
                result = enforcer.enforce(child_token, "acme_kyb", "resolve_business")
                if not result.allowed:
                    denial = result
                    break
                time.sleep(0.02)
            elapsed_ms = (time.time() - revoked_at) * 1000
            if denial is None:
                print(
                    f"trial {index}: child grant still allowed {elapsed_ms:.0f} ms after revocation",
                    file=sys.stderr,
                )
                failures += 1
                continue
            if denial.reason_code != DenialReason.GRANT_REVOKED or denial.sub_reason not in (
                RevocationSubReason.REVOKED,
                RevocationSubReason.PARENT_REVOKED,
            ):
                print(
                    f"trial {index}: unexpected denial {denial.reason_code}/{denial.sub_reason}",
                    file=sys.stderr,
                )
                failures += 1
                continue
            # The first trials warm the connection and the JWKS cache; they are
            # reported but not measured.
            if index >= WARMUP:
                latencies.append(elapsed_ms)
                revoke_calls.append(revoke_ms)
            else:
                print(f"warmup {index}: {elapsed_ms:.0f} ms")
    finally:
        enforcer.stop_revocation_feed()

    report = {
        "trials": len(latencies),
        "budget_ms": BUDGET_MS,
        "min_ms": min(latencies),
        "p50_ms": _percentile(latencies, 0.5),
        "p95_ms": _percentile(latencies, 0.95),
        "max_ms": max(latencies),
        "revoke_call_p95_ms": _percentile(revoke_calls, 0.95),
        "revoke_call_max_ms": max(revoke_calls),
        "latencies_ms": latencies,
    }
    print(f"revocation propagation (Python SDK): {json.dumps(report)}")
    if REPORT:
        with open(REPORT, "w", encoding="utf-8") as handle:
            json.dump(report, handle, indent=2)
            handle.write("\n")
    if report["p95_ms"] > BUDGET_MS:
        print(f"p95 {report['p95_ms']:.0f} ms is above the {BUDGET_MS:.0f} ms budget", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/revocation-propagation.mjs
#!/usr/bin/env node
/**
 * Measures the G-6 acceptance criterion with the TypeScript SDK: revoke a
 * parent grant and time how long a child grant keeps being authorised by
 * `enforce()` in feed mode.
 *
 * It runs as a plain Node process — the way an agent runs the SDK — against a
 * real auth service (see scripts/revocation-release-test.sh). The budget is
 * two seconds at the ninety-fifth percentile and is not configurable: that is
 * the requirement, not a setting.
 *
 *   REVOCATION_RELEASE_BASE_URL=http://127.0.0.1:3199 node scripts/revocation-propagation.mjs
 */
import { writeFileSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';

const root = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const sdkPath = join(root, 'packages', 'sdk-ts', 'dist', 'index.js');
const { Grantex, ToolManifest, Permission, DenialReason, RevocationSubReason } = await import(
  `file://${sdkPath.split('\\').join('/')}`
);

const BASE_URL = process.env.REVOCATION_RELEASE_BASE_URL;
if (!BASE_URL) {
  console.error('REVOCATION_RELEASE_BASE_URL is not set');
  process.exit(2);
}
/** The requirement (PRD G-6), deliberately not overridable. */
const BUDGET_MS = 2_000;
const TRIALS = Number(process.env.REVOCATION_RELEASE_TRIALS ?? '40');
const WARMUP = Number(process.env.REVOCATION_RELEASE_WARMUP ?? '2');
const REPORT = process.env.REVOCATION_RELEASE_REPORT;
const SCOPES = ['tool:acme_kyb:read'];

// The SDK the agent process loads must be the one in this checkout.
console.log(`SDK under test: ${sdkPath}`);

function percentile(values, fraction) {
  const sorted = [...values].sort((a, b) => a - b);
  const index = Math.min(sorted.length - 1, Math.ceil(fraction * sorted.length) - 1);
  return sorted[Math.max(0, index)];
}

async function approve(apiKey, authRequestId) {
  const response = await fetch(`${BASE_URL}/v1/authorize/${authRequestId}/approve`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', Authorization: `Bearer ${apiKey}` },
    body: '{}',
  });
  if (!response.ok) throw new Error(`approve failed: ${response.status} ${await response.text()}`);
  return (await response.json()).code;
}

const account = await Grantex.signup(
  { name: `revocation-release-${Date.now()}`, mode: 'sandbox' },
  { baseUrl: BASE_URL },
);
const admin = new Grantex({ apiKey: account.apiKey, baseUrl: BASE_URL });
const stamp = Date.now();
const root_ = await admin.agents.register({ name: `release-root-${stamp}`, scopes: SCOPES });
const middle = await admin.agents.register({ name: `release-middle-${stamp}`, scopes: SCOPES });
const leaf = await admin.agents.register({ name: `release-leaf-${stamp}`, scopes: SCOPES });

const auth = await admin.authorize({ agentId: root_.agentId, userId: `release-user-${stamp}`, scopes: SCOPES });
const code = typeof auth.code === 'string' ? auth.code : await approve(account.apiKey, auth.authRequestId);
const rootGrant = await admin.tokens.exchange({ code, agentId: root_.agentId });

const pairs = [];
for (let index = 0; index < TRIALS + WARMUP; index += 1) {
  const parent = await admin.grants.delegate({
    parentGrantToken: rootGrant.grantToken, subAgentId: middle.agentId, scopes: SCOPES, expiresIn: '1h',
  });
  const child = await admin.grants.delegate({
    parentGrantToken: parent.grantToken, subAgentId: leaf.agentId, scopes: SCOPES, expiresIn: '1h',
  });
  pairs.push({ parentGrantId: parent.grantId, childToken: child.grantToken });
}

const enforcer = new Grantex({
  apiKey: account.apiKey,
  baseUrl: BASE_URL,
  issuer: process.env.REVOCATION_RELEASE_ISSUER ?? BASE_URL,
  revocationCheck: 'feed',
  revocationFeed: { staleAfterMs: 5_000 },
});
enforcer.loadManifest(new ToolManifest({ connector: 'acme_kyb', tools: { resolve_business: Permission.READ } }));

let failures = 0;
const latencies = [];
const revokeCalls = [];
try {
  if (!await enforcer.revocationFeed().ready(10_000)) {
    console.error('the revocation feed never became fresh');
    process.exit(1);
  }
  const first = await enforcer.enforce({ grantToken: pairs[0].childToken, connector: 'acme_kyb', tool: 'resolve_business' });
  if (!first.allowed) {
    console.error(`the child grant was denied before any revocation: ${first.reason}`);
    process.exit(1);
  }

  for (const [index, pair] of pairs.entries()) {
    const started = Date.now();
    await admin.grants.revoke(pair.parentGrantId);
    // Propagation is measured from the moment the revocation is committed —
    // when the API returns — not from when the call was made. A developer on
    // the free plan is rate limited to 100 requests a minute, and the SDK
    // waits out `Retry-After`; that wait is reported separately rather than
    // charged to the feed.
    const revokedAt = Date.now();
    let denial = null;
    while (Date.now() - revokedAt < 30_000) {
      const result = await enforcer.enforce({
        grantToken: pair.childToken, connector: 'acme_kyb', tool: 'resolve_business',
      });
      if (!result.allowed) {
        denial = result;
        break;
      }
      await new Promise((resolve_) => setTimeout(resolve_, 20));
    }
    const elapsed = Date.now() - revokedAt;
    const revokeMs = revokedAt - started;
    if (revokeMs > 1_000) {
      console.log(`trial ${index}: the revoke call waited ${revokeMs} ms (plan rate limit), not counted`);
    }
    if (!denial) {
      console.error(`trial ${index}: child grant still allowed ${elapsed} ms after its parent was revoked`
        + ` (the revoke call itself took ${revokeMs} ms)`);
      if (process.env.REVOCATION_RELEASE_DIAGNOSE === '1') {
        const state = enforcer.revocationFeedState();
        const feedResponse = await fetch(`${BASE_URL}/v1/revocations?since=0&limit=1000`, {
          headers: { Authorization: `Bearer ${account.apiKey}` },
        });
        const feedBody = await feedResponse.json();
        console.error(`  client feed state: ${JSON.stringify(state)}`);
        console.error(`  server entries for this parent: ${JSON.stringify(
          feedBody.entries.filter((entry) => entry.grantId === pair.parentGrantId || entry.jti !== null).slice(0, 5),
        )}`);
        console.error(`  server entry count: ${feedBody.entries.length}, cursor ${feedBody.cursor}`);
      }
      failures += 1;
      continue;
    }
    if (denial.reasonCode !== DenialReason.GRANT_REVOKED
        || ![RevocationSubReason.REVOKED, RevocationSubReason.PARENT_REVOKED].includes(denial.subReason)) {
      console.error(`trial ${index}: unexpected denial ${denial.reasonCode}/${denial.subReason}: ${denial.reason}`);
      failures += 1;
      continue;
    }
    // The first trials warm the connection and the JWKS cache; they are
    // reported but not measured.
    if (index >= WARMUP) {
      latencies.push(elapsed);
      revokeCalls.push(revokeMs);
    } else console.log(`warmup ${index}: ${elapsed} ms (revoke call ${revokeMs} ms)`);
  }
} finally {
  await enforcer.stopRevocationFeed();
}

const report = {
  sdk: 'typescript',
  trials: latencies.length,
  warmup: WARMUP,
  budget_ms: BUDGET_MS,
  failures,
  min_ms: Math.min(...latencies),
  p50_ms: percentile(latencies, 0.5),
  p95_ms: percentile(latencies, 0.95),
  p99_ms: percentile(latencies, 0.99),
  max_ms: Math.max(...latencies),
  revoke_call_p95_ms: percentile(revokeCalls, 0.95),
  revoke_call_max_ms: Math.max(...revokeCalls),
  latencies_ms: latencies,
};
console.log(`revocation propagation (TypeScript SDK): ${JSON.stringify(report)}`);
if (REPORT) writeFileSync(REPORT, `${JSON.stringify(report, null, 2)}\n`);

if (failures > 0) {
  console.error(`${failures} trial(s) did not deny the child grant correctly`);
  process.exit(1);
}
if (report.p95_ms > BUDGET_MS || report.max_ms > 4 * BUDGET_MS) {
  console.error(`p95 ${report.p95_ms} ms / max ${report.max_ms} ms is outside the ${BUDGET_MS} ms budget`);
  process.exit(1);
}
